docs: Proposal to support multi-source apps in the UI (#17106) (#17108)
argoproj/argo-cd: docs/proposals/multiple-sources-for-applications-ui.md

  - "@ishitasequeira" 
  - "@jannfis"
  - "@rbreeze"
approvers:
  - "@jannfis"
  - "@alexmt"
  - "@crenshaw-dev"

creation-date: 2024-02-06
last-updated: 2024-02-06
---

# UI Support for Multiple Sources in Applications

This is the proposal for the UI changes to support multiple sources for an Application.

Related Issues:
* [Proposal: Support multiple sources for an application](https://github.com/argoproj/argo-cd/blob/master/docs/proposals/multiple-sources-for-applications.md)
* [Issue for the Proposal: Support multiple sources for an application](https://github.com/argoproj/argo-cd/issues/677)

## Summary

This is a follow-on proposal to supporting Multiple Sources for Applications, but for the UI.

The above [original](https://github.com/argoproj/argo-cd/blob/master/docs/proposals/multiple-sources-for-applications.md#changes-to-ui) ‘core’ proposal deferred
any design changes for the UI to a separate feature or secondary proposal. The proposal implementation that was made in [PR 10432](https://github.com/argoproj/argo-cd/pull/10432) 
enabled the UI to tolerate multi-source applications with the new Sources field, while still supporting the original Source field. 

> Here are currently the restrictions and limitations of the UI when applications with multiple sources are used:

Reviewer: ```suggestion
Here are the current restrictions and limitations of the UI when applications with multiple sources are used:
```

> 1. The application’s details page (for example [here](https://cd.apps.argoproj.io/applications/argocd/guestbook?view=tree&node=argoproj.io%2FApplication%2Fargocd%2Fguestbook%2F0&resource=)) 

Reviewer: ```suggestion
1. The application’s details page (for [example](https://cd.apps.argoproj.io/applications/argocd/guestbook?view=tree&node=argoproj.io%2FApplication%2Fargocd%2Fguestbook%2F0&resource=)) 
```

Please avoid `here` links. https://granicus.com/blog/why-click-here-links-are-bad/

currently shows one ApplicationSource, regardless of whether the application has one source or multiple sources. With the PR 10432 implementation, if the application has multiple sources,

> the UI displays only the first one from the list of sources. Also, in particular, in the Summary tab, the source parameters are non-editable. 

Reviewer: ```suggestion
the UI displays only the first of the sources. Also, in particular, in the Summary tab, the source parameters are non-editable. 
```

2. History and Rollback is disabled for multi-source applications. The button is disabled. Jorge has submitted a PR for 

> rollback which includes controller and UI changes [here](https://github.com/argoproj/argo-cd/pull/14124).

Reviewer: ```suggestion
rollback which includes [controller and UI changes](https://github.com/argoproj/argo-cd/pull/14124).
```

   <img height="40%" width="40%" src="images/history-and-rollback-button.png"/>

3. The New Application dialog currently only allows users to provide one source.

Thus, multiple source applications are not considered first class citizens in the UI.

> Note, see the Open Questions section for concerns regarding the priority or value of some of the above changes.

Reviewer: `Open Questions` should be a link

Author: Done

## Motivation

The motivation behind this change is to add a more complete story for the multiple source feature. The UI should support 
the creation of multiple source applications, and also support the viewing and editing of parameters from all sources. The three
points in the summary above are the base or core changes that need to be addressed.

### Goals

The goals of the proposal are:

> - Provide first-class support of multiple sources for applications in the UI (eg. address the aforementioned restrictions)

Reviewer: ```suggestion
- Provide first-class support of multiple sources for applications in the UI (e.g. address the aforementioned restrictions)
```

### Non-goals
* The design changes for the Argo CD CLI is beyond the scope of this proposal (The server APIs can probably be reused)

## Proposal

As mentioned in the previous summary section, the application source parameters are surfaced in the UI in three locations.
The Resource details pages, specifically, the Summary and Parameters tabs, the deployment history, and the Application
Create panel page. These pages should be updated.

### Resource Details

The following describes the current behavior and proposed changes for the Summary tab and the Parameters Tab.

#### i) Summary Tab

_Current Behavior:_ 

The current Summary tab includes source-related information, including the repository. For example, in Figure 1 below,
the REPO URL and PATH.

<img height="50%" width="50%" src="images/current-summary-tab.png"/>

Figure 1: The current Summary tab

_Proposed Change:_

To support multiple sources, the source-related information, from a single-source-based design, will be ‘pulled out’ 
and put into a new tab called **Sources**, and it will be combined with the **Parameters** tab (more details following). 

> The new **Sources** tab will allow users to view all the information related to each source, including the repo URL 

Reviewer: I'd suggest `Inputs` over `Sources`. A `parameter` is an `input` and a `source` is an `input`, but a `parameter` is not a `source`.

Author: @jsoref , thanks for the feedback.  Good suggestion.  The reason I chose Sources is that it matches the new sources field in the application manifest. Perhaps we can finalize this label in the PR, especially we can see how it will look too with the other tabs in context.

and path, chart and revision for Helm, etc. 

The view should show one source at a time (similar to what the UI is doing now, which only shows one source), but with 
widgets to allow users to cycle (via pagination or combo selector?) through each source. There are API calls to retrieve 
the data for each source.

> <img height="50%" width="50%" src="images/new-sources-tab.png"/>

Reviewer: I'm not a huge fan of this specific rendering.

- Generally this specific widget is more commonly used at the bottom of a view (sometimes both top and bottom).
- I don't have a better proposal at this time.
- Sometimes drop-downs are used for this purpose, although there are definite downsides. 

How many are likely? Just 3, or are 10, and 100 conceivable/likely?

Author: Hi @jsoref, I think there will be advantages and disadvantages.  In the appendix, I talked about collapsible cards where each card houses the source/input parameters. But it can get unwieldy if we try to support N-number of sources.

Maybe it's could end up being a combination of collapsible cards and pagination.  Note that there is some time required to load the source details.  So potentially, we can defer loading until the page is turned.  If all cards are in the one tab, then all sources need to be loaded.

Reviewer: I think I'm slightly more inclined to a vertical list of things on the left that one can click on, akin to: github.com/settings:
<img width="286" alt="image" src="https://github.com/argoproj/argo-cd/assets/2119212/67efa9a8-aa22-4e76-b2e8-bf925e235a57">

It's also more or less how IntelliJ and some other things manage...

But, I'm not wed to anything... This is more of a seeking scenarios.

Reviewer 2: I think it would be difficult to add enough information about the sources in the vertical view without making each item a bigger block with multiple fields. It might be good to have 10 sources per page and a collapsible card / pop-up modal for each source.

Author: I'd go for consistency with what we have now.  It parallels the List view like I mentioned above, and it's similar to the layout in the Settings view.  However, these cards are expandable.

Four sources, with the cards collapsed.

<img width="975" alt="FourSources-Collapsed" src="https://github.com/argoproj/argo-cd/assets/7726022/46631eab-da75-4fcf-9059-d65da5d1f0d8">

The third source is expanded.

<img width="983" alt="ThirdExpanded" src="https://github.com/argoproj/argo-cd/assets/7726022/4ce9ae9e-fbc5-43fb-ab03-d53e133aadde">

The pagination is needed in case there are those users who have many, many sources.

Reviewer: I'm really 👎  on the pagination. As long as only one source is expanded, how long / bad is the scrolling behavior?

Also, I'd rather make the URL more prominent and the kind _less_.

Author: The issue is with Helm and possibly plugin-type sources. If there are many parameters, one source card can be very long (tall).  So, in the figure above, if the first Helm source is expanded, you will have to scroll a long way down to see the other source cards. It'll be even longer if all cards are expanded.  So it can quickly get unwieldy and unusable. So this usability issue can happen even for 4 sources.

See [docs/proposals/images/helm-parameter-list.png](https://github.com/argoproj/argo-cd/pull/17108/files#diff-53ce2f5f3c4e384ad977725d2c9b02fef192d8e7201b0ab926bbf461948d4bcb)

When users visit the page, all source cards should probably be initially collapsed.

Author: The title of the card can be the Repo URL.  We can fine-tune this as part of the implementation PR?

<img width="973" alt="RepoUrlAsTitle" src="https://github.com/argoproj/argo-cd/assets/7726022/3c248dd1-bca4-4e12-a8bc-2f45a65c9599">

Reviewer: A few things:
1. the section (repo url / path / plugin source) should remain on screen while you're scrolling through its content
2. headings within it (`Directory`, `Parameters`) should be collapsible
3. ideally they should remain visible until their section scrolls off. See the GitHub Pull Request Reviews add-on for Visual Studio Code for an amazing implementation:
    https://github.com/argoproj/argo-cd/assets/2119212/df478b54-d302-4d5f-868d-ca1478d2b8d2

Reviewer: Are source numbers useful/necessary? Can someone do things with them? I suppose as anchors if I wanted to send a link to someone to look at the configuration for source n, that might be useful.

If there's no other particular value in a source number, then I probably wouldn't bother.

I'd probably be more interested in a filters/search thing so I could show only `helm` sources, or only sources from `github.com` or only ...

Author: The source numbers are not necessary.  They are not in the model too. I was thinking the source list is displayed in the order that they appear in the application manifest. I'll leave it out.

Author: > A few things:
> 
> 1. the section (repo url / path / plugin source) should remain on screen while you're scrolling through its content
> 2. headings within it (`Directory`, `Parameters`) should be collapsible
> 3. ideally they should remain visible until their section scrolls off. See the GitHub Pull Request Reviews add-on for Visual Studio Code for an amazing implementation:
>    https://github.com/argoproj/argo-cd/assets/2119212/df478b54-d302-4d5f-868d-ca1478d2b8d2

Regarding points 1 and 3 from above, yeah, the 'distiniguishing' fields should be visible so users won't get lost.  See video.

I'd like to get the READ-ONLY changes in as a first stage. (The URL and PATH fields will move to this new Sources tab from the SUMMARY tab).

The changes to support Editing the parameters will be based on this.

Note that at least the URL and Path fields will possibly be in their own EDITable section within the collapsible Source section.  The reason is that if their values change, then the Type and input parameters will change too.

Author: ![StickyTitleForSources](https://github.com/argoproj/argo-cd/assets/7726022/27a6fd19-e6a3-42b9-a760-d01903478dc7)

Here I'm taking advantage of the existing white box's additional top space that is already defined in containers.scss.

Reviewer: That's not bad -- although it's kinda hard to read. Could you give it a ring and opacity or something so it's easier to see?

Author: This is what I was referring to, where if you change the repo URL, or simply change the path within the repo, the input parameters and even the source type can change.  So it's probably better to keep these values as separate sections. So perhaps we will need two Edit buttons, like in the screenshot below? As mentioned, for single source apps, these fields are shown in the Summary section, under a different editable section.  If you change the values there, the parameters tab will be updated once you visit the tab.

<img width="980" alt="multipleEditableSections" src="https://github.com/argoproj/argo-cd/assets/7726022/66d2ca2d-374c-4d80-b481-2e4aae6ba348">

Reviewer: Hmmm... Yeah, I think it makes sense to have two edit buttons because they're really *very* unrelated things

Author: The initial PR changes for 'read-only' mode is here.  https://github.com/argoproj/argo-cd/pull/17275.  

A lot of it is refactoring. I think we can make more progress from these set of changes.

Figure 2. The new SOURCES tab will allow access to view all sources and application parameters.

#### ii) Parameters Tab
_Current Behavior:_ 

The Parameters tab shows the application parameters for the application’s repository details type or source. These can 
be Helm, Kustomize, Directory or Plugin (CMP).

_Proposed Change:_

The Parameter tab will be removed but the contents of the current parameters tab will be ‘reused’ and will be shown in 
the new **SOURCES** tab as described above. The parameters and parameter values will be shown for whatever source is 
selected by the user.

#### iii) Update/Edit Capability in the New Sources Tab

The above points describe how all the sources will be rendered.  However, the Sources tab should be the page to allow 
users to delete and add sources. (You can currently change the repo URL and path from the Summary tab, or manually edit 

> the application by hand, in the Manifest tab, but this is not considered as ‘guided’ editing).

Reviewer: ```suggestion
the application by hand, in the Manifest tab, but this is not considered as ‘guided’ editing.)
```

_Current Behavior:_ 

> The current form-based UI doesn’t support deleting a chosen/desired source of a multi-source application. It certainly 
> does not obviously support deleting the only source there is in a single-source application.

Reviewer: ```suggestion
The current form-based UI doesn’t support deleting a chosen/desired source of a multi-source application. It,
obviously, does not support deleting the only source in a single-source application.
```

_Proposed Change:_

In addition to adding the new SOURCES tab from section i) and ii), two new buttons (_Add Source_ and _Delete Source_) will 
be added to the page. For the _Add Source_ button, a separate dialog/panel will need to appear to allow the user to
input the parameters or other information.

Validation of any newly added source should prevent users from adding the same resource, and prevent users from
deleting all sources, etc.

### History and Rollback

Current Behavior: The History and Rollback button for multi-source apps is disabled. It's only enabled
for single-source apps, and shows source information as shown in Figure 3.

<img height="50%" width="50%" src="images/history-rollback-contents.png"/>

Figure 3: Source information in History

> Jorge has submitted a PR for rollback which includes controller and UI changes [here](https://github.com/argoproj/argo-cd/pull/14124). 

Reviewer: ...

Author: Removed 'here'

This can be treated as a separate, independent proposal.

> Other related changes pertain to the Last Synced Details. The Sync Details panel needs to be updated to show sync info

Reviewer 3: Info in `SYNC STATUS`  will also display multiple sources?

Author: @ashutosh16 , the SYNC STATUS header content might not have to change, but the "more information" panel probably will have to show the sources information, just like what is presented in the rollback and history.

## Proposal

As mentioned in the previous summary section, the application source parameters are surfaced in the UI in three locations.
The Resource details pages, specifically, the Summary and Parameters tabs, the deployment history, and the Application
Create panel page. These pages should be updated.

### Resource Details

The following describes the current behavior and proposed changes for the Summary tab and the Parameters Tab.

#### i) Summary Tab

_Current Behavior:_ 

The current Summary tab includes source-related information, including the repository. For example, in Figure 1 below,
the REPO URL and PATH.

<img height="50%" width="50%" src="images/current-summary-tab.png"/>

Figure 1: The current Summary tab

_Proposed Change:_

To support multiple sources, the source-related information, from a single-source-based design, will be ‘pulled out’ 
and put into a new tab called **Sources**, and it will be combined with the **Parameters** tab (more details following). 
The new **Sources** tab will allow users to view all the information related to each source, including the repo URL 
and path, chart and revision for Helm, etc. 

> The view should show one source at a time (similar to what the UI is doing now, which only shows one source), but with 
> widgets to allow users to cycle (via pagination or combo selector?) through each source. There are API calls to retrieve 
> the data for each source.

Reviewer 4: Just to clarify the proposal based on what was discussed above, does this make sense?

```suggestion
The view should show all the different sources in individual collapsible widgets with pagination to allow users to cycle through each source. 
There are API calls to retrieve the data for each source.

When there is only one source is the view, it will be expanded by default. If multiple sources are defined, they will be collapsed and the user will be able to expand them and paginate based on the number of items per page and page number. The default number of item per page should be 3 and saved in the user session.
```

Author: Regarding the comment "There are API calls to retrieve the data for each source":  I looked into this some more and I think for better performance and usability,  it should load the source details 'on demand' when the source section is expanded.  Like this:

Just shortly after clicking the expand button:
<img width="675" alt="LoadingOneSource" src="https://github.com/argoproj/argo-cd/assets/7726022/e46d3a59-e657-486e-8e92-f3055102000c">

Loading complete:
<img width="675" alt="LoadedOneSource" src="https://github.com/argoproj/argo-cd/assets/7726022/6570c815-2a80-40e6-a31b-b28042e527cc">

Reviewer 4: LGTM, I though the default could have been configurable easily. I like the async loading, I was expecting the data to be already fetched, because it is part of the Application Manifest information, but if it needs to load something, async is definitely my preference.

be the highest priority. (Before you can edit, you have to first display it.)

Here are the general enhancements to be implemented (Upstream issues to be opened if not already):

1. Create new Sources tab to replace Parameters tab so that all sources can be displayed (Read-only)
2. Update History and Rollback to show a summary of all sources of an application
   As mentioned above, this is already covered by Jorge’s [PR](https://github.com/argoproj/argo-cd/pull/14124)
3. Add _Add Source_ and _Delete Source_ buttons to Sources tab. This will depend on #1 above. (Update and Delete)
4. Update New App dialog. (Creation)
   - Support adding multiple sources in New App dialog. (This will likely depend on the Components from #1 and #3)
   - Use Sources field instead of Source field. Clean up code.

### Security Considerations
None

### Risks and Mitigations
None

### Upgrade / Downgrade Strategy
If downgraded, the UI will revert to showing just the first source.

## Drawbacks
None

## Open Questions

Supporting multiple sources in the New App dialog may not be ‘worth’ the effort? The drawback is that switching from the 

> YAML editor and form editor can lead to loss of information.

Reviewer 4: I think it is worth having an issue for tracking and if there are users of Argo where the creation of multi sources through the UI is critical for them, they can contribute and implement it.

Author: Ok, I'll open an issue for any outstanding enhancement work that could be done in the future.
